### PR TITLE
feat: Welcome message is dynamic, reflecting what the user is working on

### PR DESCRIPTION
**Pull request by dustinbyrne**
## Overview

This pull request introduces several improvements and new features to the AppMap JavaScript project, particularly focusing on the handling of Git diffs for better code reviews and welcome message functionality.

### Key Changes

1. **Git Diff and Base Commit Handling**: New functions like `getDiffLog`, `getCommitDistance`, `findNearestBranch`, and `findBaseCommit` have been added to effectively manage and calculate Git commit distances and determine the base branch. This refinement ensures that the least number of commits between the base and head are analyzed.

2. **Welcome Message Enhancements**: The welcome message now dynamically reflects the current task by analyzing changes in the current branch and within the worktree, providing relevant suggestions based on recent activities.

3. **Testing Improvements**: Unit tests for the new functionality in `navieMetadataV1` ensure coverage of different scenarios and edge cases, emphasizing the changes in the Git-related functionality.

4. **Review command enhanced**: The `@review` command will now use the new functionality to identify the base branch.

image

## Review Summary

| Aspect          | Summary                                                                                          | Considerations                                                                                  |
|-----------------|--------------------------------------------------------------------------------------------------|--------------------------------------------------------------------------------------------------|
| ✅ Correctness     | Improved handling of Git diffs and branches; base commit logic needs further verification.        | Check edge cases in `findBaseCommit` and ensure `getDiffLog` manages undefined base commits.     |
| ✅ Code Quality    | Enhanced readability with helper functions; additional comments needed in `git.ts`.               | Update inline documentation for new functions.                                                  |
| ✅ Testing         | Good test coverage; additional tests for edge cases recommended.                                 | Cover more edge cases in `getCommitDistance` and `findNearestBranch`.                            |
| ✅ Performance     | Parallel execution efficiently managed; performance impact with numerous branches requires review.| Monitor performance with large branch counts.                                                   |
| ✅ Security        | Reduced vulnerability to shell injection through `execFile` use.                                 | Maintain secure command handling practices.                                                     |
| ✅ Documentation   | Documentation improved with JSDoc; some new functions lack complete documentation.               | Add documentation for all new functions, especially in `git.ts` and `metadata.ts`.               |
| ✅ Compatibility   | Changes appear backward compatible; requires thorough testing.                                   | Conduct integration tests to confirm backward compatibility.                                    |
| ✅ Design          | Adheres to good design principles; new functions enhance overall structure.                      | Ensure ongoing adherence to design principles as codebase evolves.                               |

## Release
https://github.com/getappmap/appmap-js/pull/2138 should first be released and integrated into `appmap-services` before this branch merges. This ensures that the remote Navie provider has the capability to serve `@welcome` command requests.

### CODE REVIEW

File: packages/cli/src/lib/git.ts [hunk 1]
```diff
@@ -66,3 +67,107 @@ async function findAndSanitizeRepository(dir: string): Promise<SanitizedUri | vo
...
+async function getCommitDistance(
+  base: string,
+  head: string,
+  cwd?: string
+): Promise<number | undefined> {
+  try {
+    const result = await execute('git', ['rev-list', '--count', `${base}..${head}`], { cwd });
+    const distance = parseInt(result);
+    return Number.isNaN(distance) ? undefined : distance;
+  } catch {
```
Comment (kgilpin) [inline packages/cli/src/lib/git.ts:89]: We should probably warn here and another exceptions that aren’t logged.

File: packages/cli/src/lib/git.ts [hunk 2]
```diff
@@ -66,3 +67,107 @@ async function findAndSanitizeRepository(dir: string): Promise<SanitizedUri | vo
...
+ */
+export async function findBaseCommit(headRef = 'HEAD', cwd?: string): Promise<string | undefined> {
+  try {
+    const nearestBase = await findNearestBranch(mainBranches, headRef, cwd);
+    if (!nearestBase) return;
+
+    let baseCommit = await execute('git', ['merge-base', nearestBase, headRef], { cwd });
+    baseCommit = baseCommit.trim();
+    return baseCommit !== '' ? baseCommit : undefined;
+  } catch {
```
Comment (kgilpin) [inline packages/cli/src/lib/git.ts:137]: This one too.

File: packages/cli/src/rpc/navie/metadata.ts [hunk 3]
```diff
@@ -1,15 +1,117 @@
...
+    result.push(
+      "Based on your branch and the code changes you've made, I can tell you're working on something. I can help you answer questions about your codebase, create diagrams, plan solutions, generate code and tests, and review code changes.",
+      '',
+      'Type `@` to see a list of commands.'
+    );
+  } else {
+    result.push(
+      `It looks like you're ${welcomeSuggestion.activity}.`,
+      '',
+      'Here are some prompts you might want to try:',
```
Comment (kgilpin) [inline packages/cli/src/rpc/navie/metadata.ts:100]: The user might think of these as questions rather than prompts.

**Comment by kgilpin** [comment]
Should we also roll this out first to hosted Navie for early release, and then follow on with the wider release?

File: packages/cli/src/lib/git.ts (hunk 1 above)
Comment (dustinbyrne) [reply packages/cli/src/lib/git.ts]: I've added verbose logging for these. In this case, the exception is expected as `git` will fail on branches that don't exist. We're iterating through main, master, develop, so it's likely to occur.

**Comment by dustinbyrne** [comment]
https://github.com/getappmap/appmap-js/pull/2139/commits/de74bf453e41b232797660cbb21566821fadf2f8 ensures the dynamic welcome message is only used if Navie is using the remote provider.

**Review by kgilpin — APPROVED** [review]

**Comment by kgilpin** [comment]
Item: When I have checked out a branch, but I haven't edited any files yet, I get this message:

```
It looks like you haven't started working on a task yet, as there are no changes to your workspace or branch. To begin, you can check out a different branch or start editing files. I'll keep track of your progress, so you can return anytime for help with your task.
```

The message should reflect the fact that I have checked out a named branch, even if there are no changes yet.

So, don't instruct the user to "check out a different branch". Also, in addition to the suggestion to edit files, the user can be prompted to paste an issue description.

**Comment by kgilpin** [comment]
FYI I have updated this so that the dynamic welcome message is enabled only by environment variable, for now, as UI changes to support this seem to be very desirable (e.g. initial welcome message with a loading animation, followed by the dynamic message). 

Other improvements that are desirable:

* Customize the welcome message for a code selection.
* Load a welcome message suggestion into the chat box using a link or button.
* Suppress the `@explain` and `@diagram` command prefixes, as they are really not needed to get the desired output.
